# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 


ANSWER:

It's hard to tell which model will perform better, as both have different strengths and weaknesses. While Logistic Regression is good for binary classification (whether or not an email is spam), it is limited by only being able to model linear relationships between the different variables in the dataset, which could skew the results if any of them have falsely positive relationships. Meanwhile, random forest classifier can perform better on multi dimensional data, as is the case with this dataframe. It's hard to tell if either would be prone to overfitting/underfitting either. If I had to make a prediction, I would imagine the random forest classifier would be more accurate, as there are a ton of variables (58 columns) to work with here.

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
X = data.copy()
X = X.drop(columns='spam')

y = data['spam']

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data['spam'].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4540            0.00               0.00           0.42           0.0   
1851            0.00               0.00           0.21           0.0   
2511            0.00               0.00           0.86           0.0   
1200            0.43               0.28           0.43           0.0   
1607            0.77               0.38           0.77           0.0   
...              ...                ...            ...           ...   
4004            0.00               0.00           0.00           0.0   
3425            0.00               0.00           0.00           0.0   
3979            0.00               0.00           4.34           0.0   
2698            0.00               0.00           0.00           0.0   
317             0.00               0.00           1.31           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4540           0.00            0.00              0.00                0.00   
1851           0.42            0.00              0.00                0.00   
2511           0.00            0.00              0.00                0.00   
1200           0.04            0.11              0.04                0.21   
1607           0.00            0.57              0.00                0.57   
...             ...             ...               ...                 ...   
4004           0.00            0.00              0.00                0.00   
3425           0.00            0.00              0.00                0.00   
3979           0.00            0.00              0.00                0.00   
2698           7.14            0.00              0.00                0.00   
317            0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4540             0.00            0.00  ...                   0.0        0.000   
1851             0.21            0.00  ...                   0.0        0.000   
2511             0.00            0.00  ...                   0.0        0.000   
1200             0.40            0.69  ...                   0.0        0.000   
1607             1.15            1.15  ...                   0.0        0.027   
...               ...             ...  ...                   ...          ...   
4004             0.00            0.00  ...                   0.0        0.000   
3425             0.00            0.00  ...                   0.0        0.000   
3979             0.00            0.00  ...                   0.0        0.000   
2698             0.00            0.00  ...                   0.0        0.000   
317              0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4540        0.078        0.039        0.039        0.000        0.000   
1851        0.161        0.000        0.161        0.000        0.182   
2511        0.000        0.000        0.000        0.000        0.000   
1200        0.064        0.000        0.311        0.734        0.145   
1607        0.027        0.000        0.437        0.191        0.054   
...           ...          ...          ...          ...          ...   
4004        0.000        0.000        0.000        0.000        0.000   
3425        0.000        0.000        0.124        0.000        0.000   
3979        0.000        0.000        1.342        0.000        0.000   
2698        0.000        0.000        0.000        0.000        0.000   
317         0.000        0.000        0.000        0.199        0.000   

      capital_run_length_average  capital_run_length_longest  \
4540                       1.292                           6   
1851                       2.813                         121   
2511                       1.047                           2   
1200                       5.328                         144   
1607                      14.406              

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
X_scaler = scaler.fit(X_train)

In [8]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [15]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=42, max_iter=1500)

lr_model = logistic_regression_model.fit(X_train, y_train)

print(f"Training Data Score: {lr_model.score(X_train, y_train)}")
print(f"Testing Data Score: {lr_model.score(X_test, y_test)}")

Training Data Score: 0.9289855072463769
Testing Data Score: 0.9270199826238054


In [16]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predections = lr_model.predict(X_test)

# Review the predictions
testing_predections

array([0, 1, 1, ..., 0, 1, 0], dtype=int64)

In [17]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predections)

0.9270199826238054

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [18]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

random_forest_model = RandomForestClassifier(n_estimators=128, random_state=42)

rf_model = random_forest_model.fit(X_train, y_train)

print(f"Training Data Score: {rf_model.score(X_train, y_train)}")
print(f"Testing Data Score: {rf_model.score(X_test, y_test)}")

Training Data Score: 0.9997101449275362
Testing Data Score: 0.9539530842745438


In [19]:
# Make and save testing predictions with the saved random forest model using the test data
testing_predections_2 = rf_model.predict(X_test)

# Review the predictions
testing_predections_2

array([0, 1, 0, ..., 0, 1, 0], dtype=int64)

In [20]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predections_2)

0.9539530842745438

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

ANSWER:

--------------------------------------------------------

Logistic Regression Model:

Training Data Score: 0.9289855072463769

Testing Data Score: 0.9270199826238054

Accuracy Score: 0.9270199826238054

This model performed decently well on both training and testing datasets, and both scores are extremely close to one another, which means there is no overfitting or underfitting. The accuracy score of 92.7% means it correctly predicted spam emails 92.7% of the time.

-------------------------------------------------------

Random Forest Classifier Model:

Training Data Score: 0.9997101449275362

Testing Data Score: 0.9539530842745438

Accuracy Score: 0.9539530842745438

This model performed very well on the training data, with almost a perfect score. It correctly predicted spam emails 95.3% of the time, which is better than the logistic regression model. However, there is some considerable difference between the training and testing scores, which indicates that there may be some overfitting going on.

-------------------------------------------------------

Overall, both of these models performed very well, and while Logistic Regression had a slightly lower accuracy score than random forest, the deviation between training and testing scores in the latter indicates overfitting, which means that model performs great on this data set but would not perform as well if given other datasets to predict. So while it performed slightly worse, my choice would be the Logistic Regression model in this instance. It seems I was half-correct in my prediction, Random Forest indeed performed better. However, the overfitting makes the model less reliable than the slightly less accurate Logistic Regression.
